fcrepo4/fcrepo4
# Adds support for configurable pairtree length and count of minted identifiers

**Resolves FCREPO-2671**: https://jira.duraspace.org/browse/FCREPO-2671
**Resolves FCREPO-2672**: https://jira.duraspace.org/browse/FCREPO-2672

# What does this Pull Request do?
Adds a new configurable PID minting supplier that allows for the creation of either hierarchical or non-hierarchical identifiers.  Using the supplier:

* with no args returns a non-hierarchical identifier
* with length and count args returns a hierarchical identifier with a parent pairtree of (count) number of parts that are (length) long
* with length or count args of zero returns a non-hierarchical identifier

The length and count is optionally configurable via Spring XML, which in turn can also read values from system properties.  This resolves FCREPO-2672: https://jira.duraspace.org/browse/FCREPO-2672

# How should this be tested?
```
1. Create a resource without a slug:
$ curl -i -XPOST  -H "Link: <http://mementoweb.org/ns#OriginalResource>; rel=\"type\"" "http://localhost:8080/rest" \
-H "Content-Type: text/n3" --data-binary "<> <info:fedora/test/field> \"Versioned Resource\""

2. Check that resulting resource identifier does not contain a parent pairtree
Location: http://localhost:8080/rest/8530d479-33ec-4aa7-ab9f-36817c08aa4d

3. Uncomment the following from fcrepo-webapp/src/main/resources/spring/fcrepo-config.xml 
(and restart fcrepo-webapp if applicable.)
<bean class="org.fcrepo.kernel.api.services.functions.ConfigurableHierarchicalSupplier"
        c:desiredLength="${fcrepo.uuid.path.length:2}"
        c:desiredCount="${fcrepo.uuid.path.count:4}"/>

4. Create a resource without a slug:
$ curl -i -XPOST  -H "Link: <http://mementoweb.org/ns#OriginalResource>; rel=\"type\"" "http://localhost:8080/rest" \
-H "Content-Type: text/n3" --data-binary "<> <info:fedora/test/field> \"Versioned Resource\""

5. Check that resulting resource identifier contains a parent pairtree
Location: http://localhost:8080/rest/5c/06/12/b4/5c0612b4-265a-4333-8446-8abf02b25e75
```

Interested parties/reviewers:
@awoods @dbernstein @whikloj
